src/source-dpdk.c

```diff
@@ -0,0 +1,525 @@
+{
+#ifdef PACKET_STATISTICS
+    // Prevents access to other fields of stats structure
+    if (unlikely(ptv->queue_id >= RTE_ETHDEV_QUEUE_STAT_CNTRS))
+        return;
+
+    struct rte_eth_stats eth_stats;
+    rte_eth_stats_get(ptv->port_id, &eth_stats);
+
+    uint64_t th_pkts = StatsGetLocalCounterValue(ptv->tv, ptv->capture_dpdk_packets);
+    uint64_t th_drops = StatsGetLocalCounterValue(ptv->tv, ptv->capture_dpdk_drops);
+
+    StatsAddUI64(ptv->tv, ptv->capture_dpdk_packets, eth_stats.q_ipackets[ptv->queue_id] - th_pkts);
+    StatsAddUI64(ptv->tv, ptv->capture_dpdk_drops, eth_stats.q_errors[ptv->queue_id] - th_drops);
+    SC_ATOMIC_ADD(ptv->livedev->pkts, eth_stats.q_ipackets[ptv->queue_id] - th_pkts);
+    SC_ATOMIC_ADD(ptv->livedev->drop, eth_stats.q_errors[ptv->queue_id] - th_drops);
+#endif
+}
+
+static void DPDKReleasePacket(Packet *p)
+{
+    int retval;
+    /* Need to be in copy mode and need to detect early release
+       where Ethernet header could not be set (and pseudo packet) */
+    if (!PKT_IS_PSEUDOPKT(p) && (p->dpdk_v.copy_mode == DPDK_COPY_MODE_TAP ||
+                                        (p->dpdk_v.copy_mode == DPDK_COPY_MODE_IPS &&
+                                                !PACKET_TEST_ACTION(p, ACTION_DROP)))) {
+        retval =
+                rte_eth_tx_burst(p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+        if (unlikely(retval < 1)) {
+            retval = rte_eth_tx_burst(
+                    p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+            if (unlikely(retval < 1)) {
+                SCLogDebug("Unable to transmit the packet on port %u queue %u",
+                        p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id);
+                rte_pktmbuf_free(p->dpdk_v.mbuf);
+                p->dpdk_v.mbuf = NULL;
+            }
+        }
+    } else {
+        rte_pktmbuf_free(p->dpdk_v.mbuf);
+        p->dpdk_v.mbuf = NULL;
+    }
+
+    PacketFreeOrRelease(p);
+}
+
+/**
+ *  \brief Main DPDK reading Loop function
+ */
+TmEcode ReceiveDPDKLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+    Packet *p;
+    uint16_t nb_rx;
+    time_t last_dump = 0;
+    time_t current_time;
+
+    DPDKThreadVars *ptv = (DPDKThreadVars *)data;
+    TmSlot *s = (TmSlot *)slot;
+
+    ptv->slot = s->slot_next;
+
+    PacketPoolWait();
+    while (1) {
+        if (unlikely(suricata_ctl_flags != 0)) {
+            SCLogInfo("Stopping Suricata!");
+            DPDKDumpCounters(ptv);
+            break;
+        }
+
+        nb_rx = rte_eth_rx_burst(ptv->port_id, ptv->queue_id, ptv->received_mbufs, BURST_SIZE);
+        if (unlikely(nb_rx == 0)) {
+            continue;
+        }
+
+        ptv->pkts += (uint64_t)nb_rx;
+        for (uint16_t i = 0; i < nb_rx; i++) {
+            p = PacketGetFromQueueOrAlloc();
+            if (unlikely(p == NULL)) {
+                SCLogError(SC_ERR_MEM_ALLOC, "Failed to get Packet Buffer for DPDK mbuf!");
```

Comment:
why is this handled in such a harsh way?

Reply:
Right, it is probably too harsh. Relaxed the error handling logic and replaced it with incremented counter and `continue;`